src/util/submit-addon.js

```diff
@@ -0,0 +1,388 @@
+
+export type SignResult = {|
+  success: boolean,
+  id: string | null,
+  downloadedFiles: Array<string>,
+|};
+
+type ClientConstructorParams = {|
+  apiKey: string,
+  apiSecret: string,
+  apiUrlPrefix: string,
+  apiJwtExpiresIn?: number,
+  debugLogging?: boolean,
+  validationCheckInterval?: number,
+  validationCheckTimeout?: number,
+  approvalCheckInterval?: number,
+  approvalCheckTimeout?: number,
+  logger?: any,
+  downloadDir?: string,
+|};
+
+export default class Client {
+  apiKey: string;
+  apiSecret: string;
+  apiUrlPrefix: string;
+  apiJwtExpiresIn: number;
+  debugLogging: boolean;
+  validationCheckInterval: number;
+  validationCheckTimeout: number;
+  approvalCheckInterval: number;
+  approvalCheckTimeout: number;
+  logger: any;
+  downloadDir: string;
+
+  constructor({
+    apiKey,
+    apiSecret,
+    apiUrlPrefix,
+    apiJwtExpiresIn = 60 * 5, // 5 minutes
+    debugLogging = false,
+    validationCheckInterval = 1000,
+    validationCheckTimeout = 300000, // 5 minutes.
+    approvalCheckInterval = 1000,
+    approvalCheckTimeout = 900000, // 15 minutes.
+    logger = console,
+    downloadDir = process.cwd(),
+  }: ClientConstructorParams) {
+    this.apiKey = apiKey;
+    this.apiSecret = apiSecret;
+    this.apiUrlPrefix = `${apiUrlPrefix}addons/`;
+    this.apiJwtExpiresIn = apiJwtExpiresIn;
+    this.validationCheckInterval = validationCheckInterval;
+    this.validationCheckTimeout = validationCheckTimeout;
+    this.approvalCheckInterval = approvalCheckInterval;
+    this.approvalCheckTimeout = approvalCheckTimeout;
+    this.debugLogging = debugLogging;
+    this.logger = logger;
+    this.downloadDir = downloadDir;
+  }
+
+  fileFromSync(path: string): File {
+    return fileFromSync(path);
+  }
+
+  async doUploadSubmit(xpiPath: string, channel: string): Promise<string> {
+    const url = `${this.apiUrlPrefix}upload/`;
+    const formData = new FormData();
+    formData.set('channel', channel);
+    formData.set('upload', this.fileFromSync(xpiPath));
+    const { uuid } = await this.getJson(
+      await this.fetch(url, 'POST', formData)
+    );
+    return this.waitForValidation(uuid);
+  }
+
+  waitForValidation(uuid: string): Promise<string> {
+    let validationCheckTimeout;
+
+    return new Promise((resolve, reject) => {
+      const uploadDetailUrl = `${this.apiUrlPrefix}upload/${uuid}/`;
+      const abortTimeout = setTimeout(() => {
+        clearTimeout(validationCheckTimeout);
+
+        reject(new Error('Validation Timeout.'));
+      }, this.validationCheckTimeout);
+
+      const pollValidationStatus = async () => {
+        try {
+          const detailResponse = await this.fetch(uploadDetailUrl);
+          if (!detailResponse.ok) {
+            reject(
+              new Error(
+                `Getting upload details failed: ${detailResponse.statusText}.`
+              )
+            );
+          }
+          const detailResponseData = await detailResponse.json();
```

Comment:
Mainly it's because I did this part before I added `getJson`.  The only difference is that a non-`.ok` response would show the status here, but a generic `Bad Request` for `getJson`, but I don't think that's important.  I can change it to use `getJson` too